Add the possibility to use scripts in deps.json, for check-deps

andreriesco opened the pull request:
Add the possibility to use scripts in the deps.json file, that are
  checked by the check-deps task. Also, add the dotnet-sdk-* and the
  mono-devel install scripts.
  
  Signed-off-by: Andre Riesco <andre.riesco@toradex.com>

Commit:
cpp: Correction on the .gitignore file

Signed-off-by: Andre Riesco <andre.riesco@toradex.com>

Commit:
dotnet: Fixes .net 8 names that were with .net 6 names in some places

Signed-off-by: Andre Riesco <andre.riesco@toradex.com>

Commit:
deps: Add the possibility to use scripts in deps.json, for check-deps

Add the possibility to use scripts in the deps.json file, that are
checked by the check-deps task. Also, add the dotnet-sdk-* and the
mono-devel install scripts.

Signed-off-by: Andre Riesco <andre.riesco@toradex.com>

microhobby commented on scripts/checkDeps.ps1:
```diff
@@ -65,9 +71,22 @@ if ($_packagesToInstall.Count -gt 0) {
         foreach ($item in $_packagesToInstall) {
-            sudo apt-get install -y $item
-            
+            if ($item.EndsWith('.sh')) {
```
I would not like to force the template developer to use only shell scripts. So, my proposal for this feature would be add a new property on the `deps.json` the `installScripts`, should look like:

```json
{
    "packages": [
        "openssh-client",
        "sshpass",
        "build-essential",
        "make",
        "gdb",
        "gcc"
    ],
    "installScripts": [
        "./install-mono.sh",
        "/home/user/libs/install-lib.ps1",
        "/root/binaryname"
    ]
}
```

This new property should be an array that accepts absolute or relative to `.conf` paths.
Also, would be good to store the specific install scripts on the `.conf` folder of the specific template.

Let me know what you think.

andreriesco replied:
About the scrips, I had added just powershell and bash, but I agree it is better to have them as a separated property and support any type of script, as long as it is executed in the normal `./$scriptName` way of executing scripts

andreriesco replied:
About them being stored in each template `.conf` folder, it has the good side of being easier to understand where the script is in this templates repo. But it has the downside of maintenance, the `install-dotnet-sdk-8.sh` for example is used on 6 templates, meaning that if we change anything in this script we will need to remember to change it is all of them.
So about this point I am not so sure what would be the best option.

microhobby replied:
hmm, yeah, you have a point. What about using variables then? so, templates developers should be able to do this:

```json
{
    "packages": [
        "openssh-client",
        "sshpass",
        "build-essential",
        "make",
        "gdb",
        "gcc"
    ],
    "installScripts": [
        "./install-mono.sh",
        "${HOME}/.apollox/scripts/install-lib.ps1",
        "/root/binaryname"
    ]
}
```

The main idea here is to be able to detach the workspace from the `.apollox` folder. I think that using variables we can make easy if in some case the user want to detach from the `.apollox` folder by copying the script and changing the `deps.json`.

microhobby replied:
to start let's support only `${HOME}`, and if needed we add more ...

andreriesco replied:
I updated here the commit. I made it in a way that it supports both scripts present just in the template and also common scripts, which are copied from the `scripts/install-deps-scripts` to the `.conf/install-deps-scripts` of the project when the project is created. Also, I added it in a way that of for example the template already have this `.conf/install-deps-scripts/$script` modified to his specific use case (let's say for example that Avalonia uses a modified `install-dotnet-sdk-6.sh` for example) it gives priority to the one already present in the template.

Commit:
projectUpdater: Fix update when file on source or dest doesn't exist

Fix the update giving error when one file from the source or from the
destination doesn't exist. The output now shows a merge window with the
one that doesn't exist empty, like in a git diff when one of the files
doesn't exist

Signed-off-by: Andre Riesco <andre.riesco@toradex.com>

microhobby commented on dotnetConsole/.conf/install-deps-scripts/install-dotnet-sdk-8.sh:
```diff
@@ -0,0 +1,41 @@
+#!/bin/bash
+set -e
+
+echo "bbbbb"
```
![bunnies-what](https://github.com/toradex/vscode-torizon-templates/assets/2633321/6e76a036-8ad6-4bb6-a203-557ee0a0daec)

andreriesco replied:
Ops, residual echo from testing the update script, I will remove it here

microhobby commented on scripts/checkDeps.ps1:
```diff
@@ -40,48 +38,129 @@ if ($? -eq $false || [string]::IsNullOrEmpty($_dockerComposeV)) {
-# ask if the user want to install the packages that are not installed
-if ($_packagesToInstall.Count -gt 0) {
+# Check if there are any packages to be installed or scripts to be executed
+if ($_packagesToInstall.Count -eq 0 -and $_scriptsToInstall.Count -eq 0) {
```
optional, to be better to read:

```powershell
if (
    ($_packagesToInstall.Count -eq 0) -and
    ($_scriptsToInstall.Count -eq 0)
) {
```

microhobby commented on scripts/checkDeps.ps1:
```diff
@@ -40,48 +38,129 @@ if ($? -eq $false || [string]::IsNullOrEmpty($_dockerComposeV)) {
-if ($_packagesToInstall.Count -gt 0) {
+# Check if there are any packages to be installed or scripts to be executed
+if ($_packagesToInstall.Count -eq 0 -and $_scriptsToInstall.Count -eq 0) {
+    Write-Host -ForegroundColor DarkGreen "✅ All packages already installed and installation script executed before for this project on this machine"
```
I'm not a big fan of these large messages, could you try to keep it short?

andreriesco replied:
Sure

microhobby commented on scripts/checkDeps.ps1:
```diff
@@ -40,48 +38,129 @@ if ($? -eq $false || [string]::IsNullOrEmpty($_dockerComposeV)) {
+    Write-Host -ForegroundColor DarkGreen "✅ All packages already installed and installation script executed before for this project on this machine"
+
+    # we need to ran the check deps only if it's not ran yet
+    New-Item -Path .conf/ -Name .depok -ItemType File 2>&1 | out-null
```
This will not overwrite the scripts names added to the file?

andreriesco replied:
I tested here and it does not overwrite, but I will add a check if the file exists before running this command just to be safer.

microhobby commented on scripts/checkDeps.ps1:
```diff
@@ -40,48 +38,129 @@ if ($? -eq $false || [string]::IsNullOrEmpty($_dockerComposeV)) {
+        $_scriptsInstalledOk = $false
+
+        $_installConfirm = Read-Host `
+            "Do you want to try to install the missing debian package dependencies? <y/N>"
```
I think it would be better if we could ask just once, here we can treat it as "dependencies", scripts and packages, and then ask once and install everything.

andreriesco replied:
Sounds good, I will change it here

microhobby commented on scripts/createFromTemplate.ps1:
```diff
@@ -146,6 +146,32 @@ Copy-Item "$templateFolder/../scripts/torizonIO.ps1" "$location/.conf"
+if (($_deps.installDepsScripts.Count -gt 0)) {
+    # Create the .conf/install-deps-scripts if it doesn't exist and there are
+    # installation scripts that have the path .conf/install-deps-scripts
+    if (-not (Test-Path -Path "$location/.conf/install-deps-scripts" )){
```
Let's maintain consistency and always use camel case if need more than one word in files ors `installDepsScripts`

andreriesco replied:
Sure, I will change it here

microhobby commented on scripts/projectUpdater.ps1:
```diff
@@ -325,6 +358,19 @@ _openMergeWindow `
+# DEPS.JSON:
+_openMergeWindow `
+    $projectFolder/.conf/tmp/deps.json `
+    $projectFolder/.conf/deps.json
```
Do we really want this? IMHO would be better and more secure to only accept updates on the `deps.json` and scripts without user interaction.

andreriesco replied:
This one is mainly because of the addition of the scripts. It is very possible that the use has some script that his project depends on and he will add it there. I see this being used on a project that has more than one developer for example. Once the developer shares the project with the other he wants it to install also some dependency or installation script that his project depends on.

microhobby replied:
ok, make sense. Let's leave this way then

Commit:
deps: Fixes on commits about adding scripts to deps.json

Signed-off-by: Andre Riesco <andre.riesco@toradex.com>

microhobby commented on aspnetBlazor/.conf/deps.json:
```diff
@@ -1,7 +1,9 @@
+        "sshpass"
+    ],
+    "installDepsScripts": [
+        ".conf/install-deps-scripts/install-dotnet-sdk-8.sh"
```
these are still not in camel case

microhobby commented:
LGTM
